parseutils.c

```diff
@@ -380,44 +383,97 @@ strip_quotes(char *source, char quote)
 	{
 		char		c = *src;
 
-		if (c == quote && src[1] == '\0')
-			break;				/* skip trailing quote */
+		if (lastSlash)
+		{
+			switch (c) {
+				case '\\':
+					*dst++ = '\\';
+					src++;
+					break;
+				case 'n':
+					*dst++ = '\n';
+					src++;
+					break;
+				case 't':
+					*dst++ = '\t';
+					src++;
+					break;
+				case quote:
+					*dst++ = quote;
+					src++;
+					break;
+				default:			/* unrecognized escape just pass through; XXX: add back in the slash? */
+					*dst++ = *src++;
+					break;
+			}
+			lastSlash = false;
+		}
+		else
+		{
+			lastSlash = (c == '\\');
 
-		*dst++ = *src++;
+			if (c == quote && src[1] == '\0')
+			{
+				src++;
+				break;				/* skip trailing quote without copying */
+			}
+
+			if (lastSlash)
+				src++;
+			else
+				*dst++ = *src++;
+
+		}
 	}
 
 	*dst = '\0';
+	*source = src;
+
+	return ret;
 }
 
 /*
  * Parse tokens from a "key equals quoted value" line.
  * Examples (from Kubernetes Downward API):
- * 
+ *
  *   cluster="test-cluster1"
  *   rack="rack-22"
  *   zone="us-est-coast"
- * 
+ *   var="abc=123"
+ *   multiline="multi\nline"
+ *   quoted="{\"quoted\":\"json\"}"
+ *
  * Return two tokens; strip the quotes around the second one.
  * If exactly two tokens are not found, throw an error.
  */
 char **
 parse_keqv_line(char *line)
```

Comment:
For reference, these strings are in Go's `%q` format. It looks like the decisions for how format non-printable characters and such are made in [strconv.AppendQuote](https://pkg.go.dev/strconv#AppendQuote).

- https://github.com/kubernetes/kubernetes/blob/v1.22.0/pkg/volume/downwardapi/downwardapi.go#L262
- https://github.com/kubernetes/kubernetes/blob/v1.22.0/pkg/fieldpath/fieldpath.go#L71
- https://github.com/kubernetes/kubernetes/blob/v1.22.0/pkg/fieldpath/fieldpath.go#L36
- https://pkg.go.dev/fmt
- https://pkg.go.dev/strconv#AppendQuote

Reply:
Except for the `\x` cases, the description looks very similar to JSON.

Reply:
It is not JSON, of course. There are other escape sequences it emits. This function includes `\a`, `\U`, `\v`, and `\x` among others.

Reply:
Cool. Thanks for the source; now I can write a real parser for it! :)

Reply:
Parser written to handle this now.